src/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+
+namespace System.Net.Security
+{
+    public struct SslApplicationProtocol : IEquatable<SslApplicationProtocol>
+    {
+        private readonly byte[] _protocol;
+        private readonly ReadOnlyMemory<byte> _readOnlyProtocol;
```

Comment:
Equals on readonlymemory checks if the underlying array instance is the same, that's not the equality check that's needed here.

Reply:
I was asking why you can't implement Equals over top of `ReadOnlyMemory<byte>` rather than `byte[]`. I don't see why you need the `byte[]` field.

Reply:
I'm not aware how I can do that, I have to compare the values of the underlying byte[] for which i need to get the byte[] from readonlymemory, and looking at coreclr code, toarray() is expensive than just storing the byte[] here.

Reply:
Probably should be a `SequenceEqual` extension `ReadOnlyMemory`; meanwhile you can do
```csharp
public bool Equals(SslApplicationProtocol other)
{
    if (_readOnlyProtocol.Equals(other._readOnlyProtocol))
        return true;

    return _readOnlyProtocol.Span.SequenceEqual(other._protocol.Span);

    /*
    // Or instead of SequenceEqual
    ReadOnlySpan<byte> protocol = _readOnlyProtocol.Span; 
    ReadOnlySpan<byte> otherProtocol = other._readOnlyProtocol.Span; 
    
    if (protocol.Length != otherProtocol.Length)
        return false;
    
    for (int i = 0; i < protocol.Length; i++)
    {
        if (protocol[i] != other._protocol[i])
            return false;
    }

    return true;
    */
}
```

Reply:
something like
```c#
if(_readonlyProtocol.Span.SequenceEquals(other._readonlyProtocol.Span))
```
you could drop the readonly if you didn't need the array around as well.

Reply:
Both work, the bonus is you get to drop all the null checking because they are structs.....